\documentclass{scrartcl}
\usepackage{etoolbox}
\usepackage{bbm}
\usepackage{amsmath}
\usepackage{mathabx}
\usepackage{graphicx}
\usepackage{float}
\usepackage{parskip}
\usepackage{indentfirst}
\usepackage{subfig}
\usepackage{fancyhdr}
\pagestyle{fancy}
\setlength{\parskip}{0em}
\setlength{\parindent}{2em}

%-----------------------------------------------------------------------------
\begin{document}

%-----------------------------------------------------------------------------
% header
\lhead{Huwenbo Shi (603-778-363) shihuwenbo@ucla.edu}

% title
\newcommand*{\TitleFont}{
      \usefont{\encodingdefault}{\rmdefault}{b}{n}
      \fontsize{16}{20}
      \selectfont}
\newcommand*{\AuthorFont}{
      \usefont{\encodingdefault}{\rmdefault}{r}{n}
      \fontsize{12}{20}
      \selectfont}
\title{\TitleFont SoCal: supervised genotype calling via ellipsoid
separation for Affymetrix SNP microarray}
\author{\AuthorFont Huwenbo Shi (603-778-363) shihuwenbo@ucla.edu}
\date{}
\maketitle
%-----------------------------------------------------------------------------





%-----------------------------------------------------------------------------
% abstract
\begin{abstract}
\par
In this article, I present SoCal, a supervised genotype calling algorithm for
Affymetrix SNP microarray.
For each SNP, SoCal first efficiently identify ellipsoidal decision regions for
each genotype from reference genotype calls, and then uses these regions to
classify future SNPs into different genotypes.
Using only a small portion of training genotype calls from the HapMap Project,
SoCal achieves an accuracy of $97.5\%$ during validation.
\end{abstract}
%-----------------------------------------------------------------------------





%-----------------------------------------------------------------------------
% introduction
\section{Introduction}

\par
% broad background
Accurate genotyping of SNPs is essential to discovering true signals in
association studies.
Although next generation sequencing technology provides cheap whole--genome
sequences for genotyping SNPs, SNP microarray is still a cost--effective
genotyping technology for many specific association studies.
In an Affymetrix SNP microarray, oligonucleotide probes are used to match and
bind DNA fragments containing biallelic SNPs.
Then a fluorescence scanner scans the microarray to quantify perfect 
match and mismatch of these fragments.
Most genotype calling procedures for SNP microarray consists of two steps.
In the first step, raw information from microarray is summarized to obtain the
intensities, $\theta_A$ and $\theta_B$, of the two alleles, denoted by A and B,
of each SNP.
In the second step, SNPs are classified into genotype AA, AB, or BB based on
the allele intensities they generate.
The focus of this article is on the second step of the genotype calling
procedure---genotype calling using summarized allele intensities.

\par
% specific background
For a specific SNP, if a sample has genotype AA or BB, the allele intensity,
$\theta_A$ or $\theta_B$, will be higher respectively. 
If a sample has genotype AB, the intensities, $\theta_A$ and $\theta_B$,
will be similar.
If one plots $log(\theta_A)$ versus $log(\theta_B)$ of a SNP for a number of
samples, normally 3 ellipsoidal clusters can be observed, one for each
genotype, as shown in Figure~\ref{fig:intro_genclus}.
% state the problem
Many genotype calling algorithms use model--based unsupervised clustering
to identify these clusters and then assign genotypes to each cluster.
Although these methods are applicable to a wide range of microarrays, they
don't take advantage of genotype calls that are already available.
Also, these methods use the EM algorithm to estimate model parameters, which is
sensitive to starting parameters and slow to converge.
To utilize reference genotype calls, Rabbee et al. proposed the RLMM algorithm,
a supervised genotype calling method that uses reference genotype calls to form
decision regions by fitting bivariate Gaussian distributions on observed
allele intensities for each genotype.
These Gaussian decision regions are then used to call SNPs with unknown
genotype.
However, fitting a Gaussian distribution is known to be sensitive to outliers,
which, for SNP microarrays, can be caused by genomic structural variations.

% intro genotype cluster figure
\begin{figure}[H]
\centering
\includegraphics[scale=0.75]{intro_figs/intro_genotype_clusters.pdf}
\caption{Genotype clusters obtained from Affymetrix SNP array
allele intensity values}
\label{fig:intro_genclus}
\end{figure}

\par
% state my solution
In this article, I present SoCal, a supervised genotype calling algorithm
for Affymetrix SNP microarrays.
Instead of fitting Gaussian distributions on allele intensities with reference
genotype calls, SoCal efficiently finds ellipsoidal decision regions for each
genotype of a SNP by solving a conic programming problem.
SoCal can control the effect of outliers by assigning different weights to each
of the criteria for finding the ellipsoids---separation ratio, ellipsoid
volume, and number of enclosed points. 
After SoCal finds the ellipsoidal decision regions for each genotype of
a SNP, it uses them to call the same SNP with unknown genotype.

\par
% state the result
Using reference genotype calls from the HapMap Project as training and
validation data, SoCal achieved $99.71\%$ accuracy at a call rate of $95\%$
during leave-one-out cross-validation.
Furthermore, SoCal shows more robustness than the RLMM method when there are
outliers in the training data.
%-----------------------------------------------------------------------------





%-----------------------------------------------------------------------------
% method
\section{Methods}

% overview of method
\subsection{Overview of SoCal's genotype calling procedure}

\par
SNP allele intensities are first summarized from raw microarray data using
SNPRMA, which removes non--biological effect from the data.
After this step, SoCal calls genotypes in two steps.
In the first step, SoCal finds ellipsoidal decision regions for each of the
genotype of a SNP using reference genotype calls.
In the second step, SoCal classifies SNPs with unknown genotypes using
minimum distance classification.

% pattern separation via ellipsoids
\subsection{Pattern separation by ellipsoid}

\par
An ellipsoid $\mathcal{E} \subseteq \mathbbm{R}^{n}$ can be expressed as
$\mathcal{E} = \{x \in \mathbbm{R}^{n} | (x-c)^TE(x-c)\leqslant1\}$, where
$c$ is the center of the ellipsoid, and $E$ a positive definite matrix
denoting the shape and orientation of the ellipsoid.
Let $\{a_i\}$ be the points to be included in an ellipsoid, and $\{b_j\}$
be the points to be excluded, the problem of ellipsoidal separation is to find
$c$ and $E$ such that $(a_i-c)^TE(a_i-c)\leqslant1 \, \forall i$ and
$(b_j-c)^TE(b_j-c)>1 \, \forall j$.

% ellipsoidal pattern separation formulation
\subsection{Forming ellipsoidal decision regions for each genotype}

\par
Let $G=\{AA,AB,BB\}$ be the set of genotypes of a SNP, and $J_{AA}$, $J_{AB}$,
$J_{BB}$ the index set of samples with the corresponding genotype.
Let $X=\{(log(\theta_A),log(\theta_B))_i|i=1,\cdots,
|J_{AA}|+|J_{AB}|+|J_{BB}|\}$ be the set of log transformed allele intensities
of all the samples, and $X_{AA}=\{x_j|x_j \in X, j \in J_{AA}\}$,
$X_{AB}=\{x_j|x_j \in X, j \in J_{AB}\}$,
$X_{BB}=\{x_j|x_j \in X, j \in J_{BB}\}$ the set of log transformed allele
intensities from samples having the corresponding genotype.

\par
To find the ellipsoid that includes $X_{AA}$ and excludes
$X_{AB} \cup X_{BB}$, one
sets $\{a_i\}=X_{AA}$ and $\{b_j\}=X_{AB} \cup X_{BB}$, and
solves the following conic programming problem.
For the sake of space, detailed derivation of the problem formulation is not
presented here.
\begin{equation*}
\begin{aligned}
& \text{minimize}
&& -{\beta_{1}}k+{\beta_{2}} trace(T)+\beta_{3} \|u-\mathbbm{1}\|_1 \\  
& \text{subject to}
&& (1,a_i)^T\tilde{E}(1,a_i)\leqslant u_i \; \forall i \\
&&&(1,b_j)^T\tilde{E}(1,b_j) \ge k \; \forall j \\
&&&\tilde{E}=
    \left[
        \begin{array}{cc}
            s & v^T \\
            v & F
        \end{array}
    \right] \succeq 0 \\
&&&\left[
        \begin{array}{cc}
            F & I \\
            I & T
        \end{array}
    \right] \succeq 0
\end{aligned}
\end{equation*}

\par
In the problem formulation above $\beta_{i} > 0$ are the weights assigned to
each criteria of finding the ellipsoid---separation ratio, ellipsoid volume,
and number of enclosed points.
In SoCal, $\beta_1$, $\beta_2$, $\beta_3$ are empirically set to 1, $10^4$,
and $10^2$ respectively.

\par
Let 
$\tilde{E}^{*}=\left[
    \begin{array}{cc}
    s & v^T \\
    v & F
    \end{array}
\right]$
be the optimal solution to the problem above.
The separating ellipsoid $\mathcal{E}^{*}$ is defined as
$\mathcal{E}^{*}=
\{x \in \mathbbm{R}^{n}|(x-c^{*})^TE^{*}(x-c^{*})\leqslant2(1+k)\}$,
where $c^{*}=-F^{-1}v,\;E^{*}={{F} \over {(1-s+{c^{*}}^TF{c^{*}})}}$.

\par
To find the ellipsoid that includes $X_{AB}$ and excludes
$X_{AA} \cup X_{BB}$, one sets $\{a_i\}=X_{AB}$ and
$\{b_j\}=X_{AA} \cup X_{BB}$, and solves the above conic programming problem.
The same procedure also applies to finding the ellipsoid that includes $X_{BB}$
and excludes $X_{AA} \cup X_{AB}$.
% handling missing clusters
\subsection{Rescuing missing genotype clusters}
\par
If a SNP has moderate minor allele frequency (MAF), the genotype clusters of
that SNP are well defined, and SoCal obtains three ellipsoidal decision
regions for that SNP, one for each genotype (Figure \ref{fig:defined_cluster}).
However, if a SNP has lower MAF, some genotype cluster may not be well defined.
For these SNPs, SoCal estimates the ellipsoid for the missing genotype cluster
using the ellipsoids for the other two genotypes through simple geometric
transformations (Figure \ref{fig:sparse_cluster}).

\begin{figure}[H]
    \centering
    \subfloat[SNP with well defined genotype clusters] {
        \includegraphics[scale=0.5]{method_figs/defined_cluster.pdf}
        \label{fig:defined_cluster}
    }
    \subfloat[SNP with sparse BB genotype cluster] {
        \includegraphics[scale=0.5]{method_figs/sparse_cluster.pdf}
        \label{fig:sparse_cluster}
    }
    \caption{Each dot in the plots above represents a sample, with
             samples having HapMap reference genotype calls marked red.
             The ellipsoids were obtained using only the reference calls.}
    \label{fig:defined_sparse_cluster}
\end{figure}

% missing genotype aa cluster
\subsubsection{Missing genotype AA or BB cluster}

\par
If genotype $AA$ cluster of a SNP has less than 3 reference calls, SoCal
first finds the ellipsoids for genotype $AB$ and $BB$ clusters, and then
estimates that for genotype $AA$ cluster through simple geometric
transformations.

\par
Let $\mathcal{E}_{AB}=
\{x \in \mathbbm{R}^{n}|(x-c_{AB})^TE_{AB}(x-c_{AB}) \leqslant 1\}$
and $\mathcal{E}_{BB}=
\{x \in \mathbbm{R}^{n}|(x-c_{BB})^TE_{BB}(x-c_{BB}) \leqslant 1\}$
be the ellipsoids obtained for genotype $AB$ and $BB$ clusters,
and $n_{AB}$, $n_{BB}$ the unit vectors pointing in the direction of
the major axis of the corresponding ellipsoid.
SoCal estimates the center of $\mathcal{E}_{AA}$, the ellipsoid for genotype
$AA$ cluster, by reflecting $c_{BB}$, the center of $\mathcal{E}_{BB}$, across
the major axis of $\mathcal{E}_{AB}$.
To estimate the orientation of $\mathcal{E}_{AA}$, SoCal first determines the
angle between $n_{AB}$ and $n_{BB}$, and then applies a rotation matrix of
that angle on $E_{AB}$.

\par
Formally, let $\mathcal{E}_{AA}=
\{x \in \mathbbm{R}^{n}|(x-c_{AA})^TE_{AA}(x-c_{AA}) \leqslant 1\}$ be the
estimated ellipsoid for genotype $AA$ cluster, and $\alpha$ the angle between
$n_{AB}$ and $n_{BB}$, then
$c_{AA}=-c_{BB}+2c_{AB}+2n_{AB}((c_{BB}-c_{AB})^{T}n_{AB})$, and
$E_{AA}=R^{T}E_{AB}R$, where $R$ is a rotation matrix of angle $\alpha$.

\par
If genotype $BB$ cluster is missing, the center and orientation of the
ellipsoid for that cluster is estimated in a similar way.
Formally, let $\mathcal{E}_{BB}=
\{x \in \mathbbm{R}^{n}|(x-c_{BB})^TE_{BB}(x-c_{BB}) \leqslant 1\}$ be the
estimated ellipsoid for genotype $BB$ cluster, and $\alpha$ the angle between
$n_{AB}$ and $n_{AA}$, then
$c_{BB}=-c_{AA}+2c_{AB}+2n_{AB}((c_{AA}-c_{AB})^{T}n_{AB})$, and
$E_{BB}=R^{T}E_{AB}R$, where $R$ is a rotation matrix of angle $-\alpha$.

% missing genotype ab cluster
\subsubsection{Missing genotype AB cluster}

Although SNPs with genotype $AB$ cluster missing were not observed in HapMap
reference genotype calls, for completeness, for these SNPs SoCal first
obtains, $\mathcal{E}_{AA}$ and $\mathcal{E}_{BB}$,
the ellipsoids for genotype $AA$ and $BB$ cluster, and then estimates the
center of $\mathcal{E}_{AB}$, the ellipsoid for the missing cluster, using
the mid-point between the centers of $\mathcal{E}_{AA}$ and $\mathcal{E}_{BB}$.
The orientation of $\mathcal{E}_{AB}$ is obtained by applying a rotation
to the ellipsoid with the minimum volume among $\mathcal{E}_{AA}$ and
$\mathcal{E}_{BB}$.

\par
Formally, let $\mathcal{E}_{AB}=
\{x \in \mathbbm{R}^{n}|(x-c_{AB})^TE_{AB}(x-c_{AB}) \leqslant 1\}$ be the
estimated ellipsoid for genotype $AB$ cluster, and $\alpha$ the angle between
$n_{AA}$ and $n_{BB}$, then
$c_{AB}=(c_{AA}+c_{BB})/2$, and
$E_{AB}=R^{T}\hat{E}R$, where $\hat{E}$ is the matrix of the ellipsoid
with the minimum volumne among $\mathcal{E}_{AA}$ and $\mathcal{E}_{BB}$, and
$R$ a rotation matrix of angle $\pm\alpha/2$.
The sign of the angle of rotation is dependent on the choise of ellipsoid on
which rotation is applied---positive for $\mathcal{E}_{AA}$ and negative
for $\mathcal{E}_{BB}$.

% genotype calling and outlier detection
\subsection{Genotype calling}

\par
After the ellipsoidal decision regions,
$\mathcal{E}_{g}=
\{x \in \mathbbm{R}^{n} | (x-c_g)^TE_g(x-c_g) \leqslant 1\},
\forall g\in\{AA,AB,BB\}$ of a SNP are obtained, SoCal uses them to classify
SNPs with unknown genotypes using minimum distance classification.

\par
If a sample has allele intensity $\theta_A$ and $\theta_B$ at SNP $n$,
SoCal first computes
$D_g=\sqrt{(x-c_g)^TE_g(x-c_g)}$,
where $x=(log(\theta_A), log(\theta_B))$, for each $g\in\{AA,AB,BB\}$.
SoCal then calls the genotype, $\mathcal{G}$, of that sample at SNP $n$
as the genotype having the minimum $D_g$, that is,
$\mathcal{G}=\operatorname*{arg\,min}_{g\in \{AA,AB,BB\}}D_g$.
SoCal defines $\lambda=1-D_{\mathcal{G}}/(D_{AA}+D_{AB}+D_{BB})$ to quantify
the confidence of each genotype call. 
%-----------------------------------------------------------------------------





%-----------------------------------------------------------------------------
% data
\section{Materials}

\par
The microarray used for evaluation in this project was the Affymetrix GeneChip
Human Mapping 50K Xba Array, which contains 58,960 SNPs.
Raw microarray data for 270 samples was obtained from HapMap FTP.
And reference genotype calls were obtained from HapMap using HapMart.

\par
After removing strand-ambiguous SNPs and SNPs not present on HapMart from
the microarry, 16,387 SNPs were left.
Figure~\ref{fig:data_maf_dist} shows the minor allele frequency distribution
for these SNPs.

% maf distribution figure
\begin{figure}[H]
\centering
\includegraphics[scale=0.75]
{data_figs/maf_dist.pdf}
\caption{Minor allele frequency distribution for the 16,387 SNPs.}
\label{fig:data_maf_dist}
\end{figure}


\par
From the 16,387 SNPs, 4,064 SNPs with two genotype clusters having less than
3 reference genotype calls were further removed from the microarry.
Among these SNPs, 3,596 are monomorphic SNPs.
In total, 12,323 SNPs were left for evaluation.
On average, each of these SNPs has 83 reference genotype calls.
%-----------------------------------------------------------------------------





%-----------------------------------------------------------------------------
% result
\section{Results}

% result for comparing socal with hapmap calls
\subsection{Cross-validation with HapMap reference calls}

\par
To evaluate the accuracy of SoCal, I compared the genotype calls made by SoCal
with the reference calls from HapMap through leave-one-out
cross-validation.
For each SNP, I used one sample from the reference set as validation data and
the rest as training data.
I repeated this process until all the samples in the reference set were used
as validation data exactly once.

\par
Table~\ref{table:cmp_hapmap_socal_100} shows the genotype calls made by SoCal
and those available on HapMap during leave-one-out cross-validation.
At a call rate of 100\%, SoCal made 1,081,319 calls in total, out of which
1,069,857 were concordant with HapMap calls, achieving a concordance
rate of 98.94\%.

% 100% call rate result
\begin{table}[H]
\centering
\begin{tabular}{l*{5}{r}r}
    \hline
    HapMap/SoCal  & AA       & AB      & BB      & No Call \\ \hline
    AA            & 360,289  & 2,282   & 1,058   & 0  \\
    AB            & 2,667    & 341,012 & 2,257   & 0  \\
    BB            & 851      & 2,347   & 368,556 & 0  \\ \hline
\end{tabular}
\caption{At a call rate of 100\%, SoCal achieved 98.94\% concordance rate
in the leave-one-out cross-validation with HapMap reference calls.}
\label{table:cmp_hapmap_socal_100}
\end{table}

\par
By increasing the threshold for the confidence score $\lambda$,
SoCal can achieve higher concordance rate at the cost of making fewer calls.
At $\lambda\geqslant0.863$ and call rate of 95\%, SoCal made 1,028,258 calls in
total, out of which 1,025,242 were concordant with HapMap calls, achieving a
concordance rate of 99.71\%.
Table~\ref{table:cmp_hapmap_socal_95} shows the detailed comparison between
SoCal and HapMap calls at a call rate of 95\%.

% 95% call rate result
\begin{table}[H]
\centering
\begin{tabular}{l*{5}{r}r}
    \hline
    HapMap/SoCal  & AA       & AB      & BB      & No Call \\ \hline
    AA            & 348,221  & 390     & 298     & 14,720  \\
    AB            & 710      & 319,394 & 775     & 25,057  \\
    BB            & 410      & 427     & 357,627 & 13,290  \\ \hline
\end{tabular}
\caption{At a call rate of 95\%, SoCal achieved 99.71\% concordance rate
in the leave-one-out cross-validation with HapMap reference calls.}
\label{table:cmp_hapmap_socal_95}
\end{table}

\par
Figure~\ref{fig:result_sh_crvacc} shows the cordance rate of SoCal by varying
the call rate.
At a fixed high concordance rate, SoCal can achieve very high call rate.
This result is comparable to the results reported by many state-of-the-art
genotype calling methods.

% accuracy as function of call rate
\begin{figure}[H]
\centering
\includegraphics[scale=0.75]
{result_figs/cmp_socal_hapmap/cmp_socal_hapmap_cr_vs_acc.pdf}
\caption{Concordance rate of SoCal in the leave-one-out cross-validation with
HapMap reference calls as a function of call rate.}
\label{fig:result_sh_crvacc}
\end{figure}

% comparison with rlmm
\subsection{Comparison with RLMM in the presence of outliers}

\par
I investigated how robust SoCal is when there are outliers in the
training data.
For comparison, I implemented the RLMM algorithm, which independently
fits bivariate Gaussian distributions on the log transformed allele intensities
of each genotype cluster and then classifies SNPs with unknown genotype into
the distribution having minimum Mahalanobis distance. 

\par
For fair and accurate comparison, I selected a subset of 3,442 SNPs that have
more than 10 reference calls for each of the genotype cluster from the set of
12,323 SNPs used for evaluation in the previous section.
To simulate outliers, for each SNP, I first estimated $\mu_g$, the mean of
the log transformed allele intensities of each genotype cluster, and then
drew one outlier for each genotype cluster from the Gaussian distribution
$N(\mu_g, \gamma I)$, where $I$ is the identity matrix and $\gamma$ a
positive constant controlling the variance of the distribution.
By increasing $\gamma$, one increases the effect of outliers.

\par
Figure~\ref{fig:defined_sparse_cluster} shows the ellipsoidal decision regions
obtained by SoCal and the level curves of the Gaussian decision regions
obtained by RLMM before and after an outlier is introduced to the genotype AA
cluster.
Before an outlier is introduced, both SoCal and RLMM can make accurate
genotype calls.
However, after an outlier is introduced into the genotype AA cluster, the
variance of the Gaussian decision region obtained by RLMM is significantly
affected, making the classificatoin much less accurate.
On the other hand, because SoCal not only puts penalty on outliers but also
jointly uses data from other genotype clusters to form the decision region,
it can still classify most SNPs into correct genotypes.

% figure to illustrate socal vs gauss in terms of noise
\begin{figure}[H]
    \centering
    \subfloat[Decision regions found by SoCal when there is no outlier] {
        \includegraphics[scale=0.5]
            {result_figs/cmp_socal_gauss_noise_shape/ellipse_no_noise.pdf}
        \label{fig:socal_region_no_noise}
    }
    \subfloat[Decision regions found by RLMM when there is no outlier] {
        \includegraphics[scale=0.5]
            {result_figs/cmp_socal_gauss_noise_shape/gauss_no_noise.pdf}
        \label{fig:gauss_region_no_noise}
    }
    \\
    \subfloat[Decision regions found by SoCal when there is no outlier] {
        \includegraphics[scale=0.5]
            {result_figs/cmp_socal_gauss_noise_shape/ellipse_noise.pdf}
        \label{fig:socal_region_no_noise}
    }
    \subfloat[Decision regions found by RLMM when there is no outlier] {
        \includegraphics[scale=0.5]
            {result_figs/cmp_socal_gauss_noise_shape/gauss_noise.pdf}
        \label{fig:gauss_region_no_noise}
    }
    \caption{Each dot in the plots above represents a sample, with
             samples having HapMap reference genotype calls marked red.
             The ellipsoids were obtained using only the reference calls.}
    \label{fig:defined_sparse_cluster}
\end{figure}

% accuracy as function of noise variance
\begin{figure}[H]
\centering
\includegraphics[scale=0.75]
{result_figs/cmp_socal_gauss_noise/socal_gauss_cmp_noise.pdf}
\caption{Concordance rate of SoCal and RLMM in the leave-one-out
cross-validation with HapMap reference calls as a function of noise variance.}
\label{fig:result_cmp_noise}
\end{figure}


\subsection{Comparison with CRLMM calls}
TODO: Compare with CRLMM
concordance rate
call rate

%-----------------------------------------------------------------------------






%-----------------------------------------------------------------------------
% discussion
\section{Discussion}

\par
SoCal demonstrates the possibility of using ellipsoids as decision
regions for classifying SNPs into genotypes.
In experiments involving HapMap data, SoCal is able to achieve $97.5\%$
genotype calling accuracy using only a small portion of reference genotype
calls.

\par
SoCal is different from many other genotype calling algorithms in different
aspects.
First, SoCal is a supervised genotype caller that uses reference genotype calls
to form decisions regions and then call genotypes for future SNPs.
Second, SoCal is efficient - the problem of finding separating ellipsoids
can be formulated as a conic programming problem and solved in polynomial time,
and is guaranteed to find a global optimum.
Third, the decision regions of SoCal can be used repeatedly.
Once the decision regions of a SNP for a particular SNP array are found,
future genotype calls can repeatedly use these regions without having to
first train a genotype caller.

\par
SoCal can be improved and extended in many ways.
First, for a SNP with low minor allele frequencies, the ellipsoid learned 
directly from one genotype may not be the best ellipsoid for classification
because of limited amount of data.
Instead, one can use the ellipsoids learned for other genotypes, which have
more data, to infer the ellipsoid for the genotype with relatively less data.
Second, SoCal can be extended to detect copy number variations in samples.
Individual samples with copy number variations may have extra A's or B's at
a SNP location. And if one plots the allele intensities $(I_A,I_B)$ for these
individuals, these data point will appear as outliers.
SoCal can use the ellipsoids learned from reference genotype calls to detect
outliers and then call copy number variations accordingly.

\par
In summary, SoCal is an efficient and accurate genotype caller for Affymetrix
SNP arrays. It can also be extended and improved to have more and
better functionalities.

%-----------------------------------------------------------------------------




%-----------------------------------------------------------------------------
% references

\begin{thebibliography}{9}
\bibitem{rho2010}
Rho, S. W., Abell, G. C., Kim, K., Nam, Y., \& Bae, J. (2010).
Comparing microarrays and next-generation sequencing technologies for microbial
ecology research. Trends in Biotechnology, 28, 291-299.
\bibitem{norlen2008}
Norl\'{e}n, H., Pettersson, E., Ahmadian, A., Lundeberg, J., \& Sundberg, R.
(2008). Classification of SNP genotypes by a Gaussian mixture model in
competitive enzymatic assays.
Mathematical Statistics Stockholm University Research Report, 3, 1-26.
\bibitem{lin2008}
Lin, Y., Tseng, G. C., Cheong, S. Y., Bean, L. J., Sherman, S. L., \&
Feingold, E. (2008). Smarter clustering methods for SNP genotype calling.
Bioinformatics, 24, 2665-2671.
\bibitem{wu1983}
Wu, C. F. (1983). On the Convergence Properties of the EM Algorithm.
The Annals of Statistics, 11, 95-103.
\bibitem{rabbee2005}
Rabbee, N., \& Speed, T. P. (2005).
A genotype calling algorithm for affymetrix SNP arrays.
Bioinformatics, 22, 7-12.
\bibitem{zhou2014}
Zhou, X., \& Stephens, M. (2014).
Efficient multivariate linear mixed model algorithms for genome-wide
association studies.
Nature Methods, 11, 407-409.
\bibitem{glineur1998}
Glineur F. (1998). Pattern separation via ellipsoids and conic programming.
(MS Thesis). Faculté Polytechnique de Mons, Mons, Belgium.
\end{thebibliography}

%-----------------------------------------------------------------------------

\end{document}
